### ISSUE TITLE
Ear Training doesn't always recognize notes

### ISSUE THREAD

**Issue by HansHermann123** [leafo/sightreading.training#34]
When I use the ear training, correctly played phrases often don't get recognized, especially when played legato. If I play a really short staccato it often works, but if I play disconnected notes in a normal tempo they are not recognized.

**Comment by leafo** [comment]
It's required to play notes completely separately, are you sure your fingers aren't slightly depressing keys between notes? In any case, I think the detection can be improved to be less strict about overlapping notes.

**Comment by HansHermann123** [comment]
Yes I played the notes perfectly separated. I guess the problem could be  
that my e-piano constantly sends timecode data (Midi Statuscode F8) which  
probably screws up the separated note detection. Unfortunately it can't be  
turned off on the e-piano. Don't know if a filter is available that lets  
only notes pass through?

Cheers, Thorsten


Am .06.2018, 19:09 Uhr, schrieb leaf <notifications@github.com>:

>
> It's required to play notes completely separately, are you sure your  
> fingers aren't slightly depressing keys between notes? In any case, I  
> think the detection >can be improved to be less strict about overlapping  
> notes.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.



-- 
Erstellt mit Operas E-Mail-Modul: http://www.opera.com/mail/

**Comment by leafo** [comment]
Thanks for the info. Only note on and off events should be processed. I think it might be worthwhile to add a midi event recording tool so I can have people with bugs capture the events and send them over for me to replay.

**Comment by leafo** [comment]
Also, just for confirmation, this is with the "melody playback" ear training exercise right?

**Comment by HansHermann123** [comment]
Exactly

**Comment by HansHermann123** [comment]
So yeah, the midi timecode stuff shouldn't be the problem because the "Staff" playing works perfectly

**Comment by fregren** [comment]
I get this problem as well. For me the bug is because the melody is listed using flats, but the keyboard gives sharps and the notes are compared using strings. For example the melody wants "Db" but when I press that on the keyboard it's logged as "C#" which is from a string perspective different.

So perhaps alter MelodyPlaybackExercise.checkForMatch() to not use string comparison, or make sure that all notes are only ever represented in one way. A more elaborate solution might be to store the notes as integer values instead of strings, perhaps the MIDI note value, and then it becomes an integer comparison instead.

https://github.com/leafo/sightreading.training/blob/master/static/js/components/ear_training/melody_playback_exercise.es6#L97
